File: 2021/06/08/main.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func buildTree(preorder []int, inorder []int) *TreeNode {
	d := len(inorder)
	mp := make(map[int]int)
	for i := 0; i < d; i++ {
		mp[inorder[i]] = i
	}
	res, _ := toTree(preorder, mp, 0, 0, d-1)
	return res
}

func toTree(preorder []int, mp map[int]int, index, start, end int) (*TreeNode, int) {
	if start > end {
		return nil, index
	}
	cur := preorder[index]
	treeNode := &TreeNode{Val: cur}
	index += 1

	treeNode.Left, index = toTree(preorder, mp, index, start, mp[cur]-1)
	treeNode.Right, index = toTree(preorder, mp, index, mp[cur]+1, end)

	return treeNode, index
}

File: 2021/06/08/main.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        mp = {o: i for i, o in enumerate(inorder)}
        def toTree(index, start, end):
            if start > end:
                return None, index
            
            treeNode = TreeNode(preorder[index])
            index += 1
            idx = mp[treeNode.val]
            
            treeNode.left, index = toTree(index, start, idx - 1)
            treeNode.right, index = toTree(index, idx + 1, end)

            return treeNode, index
        res, index = toTree(0, 0, len(inorder) - 1)
        return res
